dockerfile-prs

```diff
@@ -135,52 +148,61 @@ EXPOSE 8080
 
-RUN --mount=type=cache,id=pnpm,target=/pnpm/store \
-    pnpm install --frozen-lockfile --prefer-offline
+COPY pnpm-lock.yaml pnpm-workspace.yaml package.json ./
+COPY packages/common/package.json packages/common/
+COPY packages/warehouses/package.json packages/warehouses/
+COPY packages/backend/package.json packages/backend/
+COPY packages/frontend/package.json packages/frontend/
 
-# Install Sentry CLI if environment variables are set
-ARG SENTRY_AUTH_TOKEN=""
-ARG SENTRY_ORG=""
-ARG SENTRY_RELEASE_VERSION=""
-ARG SENTRY_FRONTEND_PROJECT=""
-ARG SENTRY_BACKEND_PROJECT=""
-ARG SENTRY_ENVIRONMENT=""
+# Produce a node_modules graph cache, reusable by all build stages
+RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm fetch
 
-RUN if [ -n "${SENTRY_AUTH_TOKEN}" ] && [ -n "${SENTRY_ORG}" ] && [ -n "${SENTRY_RELEASE_VERSION}" ]; then \
-    npm install -g @sentry/cli; \
-    fi
+RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install --frozen-lockfile --offline
 
-# Build common
-COPY packages/common/tsconfig.json ./packages/common/
-COPY packages/common/tsconfig.build.json ./packages/common/
-COPY packages/common/tsconfig.esm.json ./packages/common/
-COPY packages/common/tsconfig.cjs.json ./packages/common/
-COPY packages/common/tsconfig.types.json ./packages/common/
-COPY packages/common/src/ ./packages/common/src/
+# Increase Node.js heap size for TypeScript compilation during parallel builds
+ENV NODE_OPTIONS="--max-old-space-size=4096"
+
+############################
+# Build: common
+############################
+FROM prod-builder AS build-common
+COPY packages/common/tsconfig*.json packages/common/
+COPY packages/common/src/ packages/common/src/
 RUN pnpm -F @lightdash/common build
+# output: /usr/app/packages/common/dist
+
+############################
+# Build: warehouses
+############################
+FROM prod-builder AS build-warehouses
 
-# Build warehouses
-COPY packages/warehouses/tsconfig.json ./packages/warehouses/
-COPY packages/warehouses/src/ ./packages/warehouses/src/
+COPY --from=build-common /usr/app/packages/common/ packages/common/
+
+COPY packages/warehouses/tsconfig.json packages/warehouses/
+COPY packages/warehouses/src/ packages/warehouses/src/
 RUN pnpm -F @lightdash/warehouses build
+# output: /usr/app/packages/warehouses/dist
+
+############################
+# Build: backend
+############################
+FROM prod-builder AS build-backend
+COPY --from=build-common /usr/app/packages/common/ packages/common/
+COPY --from=build-warehouses /usr/app/packages/warehouses/ packages/warehouses/
 
-# Build backend
-COPY packages/backend/tsconfig.json ./packages/backend/
-COPY packages/backend/tsconfig.sentry.json ./packages/backend/
-COPY packages/backend/src/ ./packages/backend/src
+COPY packages/backend/tsconfig*.json packages/backend/
+COPY packages/backend/src/ packages/backend/src/
 
-# Conditionally build backend with sourcemaps if Sentry environment variables are set
+ARG SENTRY_AUTH_TOKEN=""
```

Comment:
why do we need to specify this sentry args on all the layers ?

Reply:
I'll consolidate those in the base layer